### ISSUE TITLE
Rule Change: no-restricted-globals should detect restricted globals accessed via global object

### ISSUE THREAD

**Issue by sethamus** [eslint/eslint#19804]
### What rule do you want to change?

no-restricted-globals

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
/*eslint no-restricted-globals: ["error", "Promise"]*/

const fetchData = window.Promise.resolve("data");
```

### What does the rule currently do for this code?

The rule currently reports no errors.

### What will the rule do after it's changed?

The rule will report an error for `Promise`.

### Participation

- [x] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Generally makes sense to me. As discussed in https://github.com/eslint/eslint/pull/19781#discussion_r2112068401, it might be better to put this behind an option.

The question is - how will the rule know which global(s) represent the global object?

**Comment by mdjermanovic** [comment]
@eslint/eslint-team thoughts about this?

**Comment by fasttime** [comment]
> Generally makes sense to me. As discussed in [#19781 (comment)](https://github.com/eslint/eslint/pull/19781#discussion_r2112068401), it might be better to put this behind an option.

I also think this could be useful.

> The question is - how will the rule know which global(s) represent the global object?

The globals are for example `window` and `self` for browsers, or `global` and `root` for Node.js and alike, additionally to `globalThis` for all new environments. Maybe there could be a single option to configure the names of global variables that represent the global object in the program, defaulting to an empty value.

**Comment by nzakas** [comment]
I like adding this as an option (off by default for backwards compatibility).

We could provide an array where people can add global names in addition to `globalThis`. I can't think of a way to detect the environment and then figure out globals, although the list of global is small and known so we could, in theory, just detect them all automatically.

**Comment by JoshuaKGoldberg** [comment]
+1 on automatic global detection. It's very unusual and weird to use anything not mentioned in https://github.com/eslint/eslint/issues/19804#issuecomment-2939686885. Scope analysis can show whether the potential global is defined locally. Any rare + weird edge case would also be detected by `no-undef` (for JS users) or type checking (for TS users). 

If someone has a use case for a non-standard global, I think that would be best as a followup issue & discussion.

**Comment by nzakas** [comment]
> +1 on automatic global detection.

@JoshuaKGoldberg can you explain what you mean by this? Do you mean we should keep a list of known globals in the rule and loop over the global scope to see which are present? 

Marking as accepted as there's consensus on moving forward.

**Comment by JoshuaKGoldberg** [comment]
> Do you mean we should keep a list of known globals in the rule and loop over the global scope to see which are present?

Exactly, that.

**Comment by nzakas** [comment]
I think that will work.

I also think we'd still need an option to specify additional global object names. While there are a finite set of known names, that doesn't mean those are the only names ever used in any JavaScript runtime.
